dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorC11.java

```diff
@@ -118,12 +120,24 @@ public List<Event> visitAtomicXchg(AtomicXchg e) {
         Load load = newRMWLoadWithMo(e.getResultRegister(), address, Tag.C11.loadMO(mo));
         RMWStore store = newRMWStoreWithMo(load, address, e.getValue(), Tag.C11.storeMO(mo));
 
-        return tagList(eventSequence(
+        return tagList(e, eventSequence(
                 load,
                 store
         ));
     }
 
+    @Override
+    public List<Event> visitControlBarrier(ControlBarrier e) {
+        Event entryFence = EventFactory.newControlBarrier(e.getName() + "_entry", e.getId());
+        entryFence.addTags(Tag.OpenCL.ENTRY_FENCE, C11.MO_RELEASE);
+        Event exitFence = EventFactory.newControlBarrier(e.getName() + "_exit", e.getId());
+        exitFence.addTags(Tag.OpenCL.EXIT_FENCE, C11.MO_ACQUIRE);
+        return tagList(e, eventSequence(
+                entryFence,
+                exitFence
+        ));
+    }
+
```

Comment:
What is the point of having this split into an entry and exit barrier? Wouldn't we get the same behavior by having a single barrier tagged both as entry/exit and rel/acq?

Reply:
Even the natural language spec talks about two fences, so better to keep it like this.